### ISSUE TITLE
Error when creating provisioner - failed calling webhook

### ISSUE THREAD

**Issue by midestefanis** [aws/karpenter#1165]
## Version

**Karpenter**: v0.5.4

**Kubernetes**: v1.21.5

## Actual Behavior

Can't install provisioner. When I try to apply the provisioner I get this error:

`Error from server (InternalError): error when creating "provisioner.yaml": Internal error occurred: failed calling webhook "defaulting.webhook.provisioners.karpenter.sh": Post "https://karpenter-webhook.karpenter.svc:443/default-resource?timeout=10s": context deadline exceeded
`

I don't have ERROR logs neither in the pod of the controller like the one of the webhook.

## Steps to Reproduce the Problem

I am following this documentation -> https://karpenter.sh/docs/getting-started-with-terraform/

## Resource Specs and Logs
<!-- Include Provisioner spec(s), pod spec(s), and Karpenter controll

[karpenter-webhook-7fc64bdc8f-r664x.log](https://github.com/aws/karpenter/files/7889529/karpenter-webhook-7fc64bdc8f-r664x.log)
[karpenter-controller-6f6cb4d858-fc9p5.log](https://github.com/aws/karpenter/files/7889530/karpenter-controller-6f6cb4d858-fc9p5.log)

**Comment by midestefanis** [comment]
@mikesir87 Is running. I leave controller and webhook logs:

**Controller** 

``` 
2022-01-18T13:41:08.227Z	INFO	Successfully created the logger.
2022-01-18T13:41:08.227Z	INFO	Logging level set to: info
{"level":"info","ts":1642513268.3330882,"logger":"fallback","caller":"injection/injection.go:61","msg":"Starting informers..."}
I0118 13:41:08.377955       1 leaderelection.go:243] attempting to acquire leader lease karpenter/karpenter-leader-election...
2022-01-18T13:41:08.378Z	INFO	controller	starting metrics server	{"commit": "7e79a67", "path": "/metrics"}
I0118 13:41:26.328468       1 leaderelection.go:253] successfully acquired lease karpenter/karpenter-leader-election
2022-01-18T13:41:26.328Z	INFO	controller.controller.counter	Starting EventSource	{"commit": "7e79a67", "reconciler group": "karpenter.sh", "reconciler kind": "Provisioner", "source": "kind source: /, Kind="}
2022-01-18T13:41:26.328Z	INFO	controller.controller.counter	Starting EventSource	{"commit": "7e79a67", "reconciler group": "karpenter.sh", "reconciler kind": "Provisioner", "source": "kind source: /, Kind="}
2022-01-18T13:41:26.328Z	INFO	controller.controller.counter	Starting Controller	{"commit": "7e79a67", "reconciler group": "karpenter.sh", "reconciler kind": "Provisioner"}
2022-01-18T13:41:26.329Z	INFO	controller.controller.provisioning	Starting EventSource	{"commit": "7e79a67", "reconciler group": "karpenter.sh", "reconciler kind": "Provisioner", "source": "kind source: /, Kind="}
2022-01-18T13:41:26.329Z	INFO	controller.controller.provisioning	Starting Controller	{"commit": "7e79a67", "reconciler group": "karpenter.sh", "reconciler kind": "Provisioner"}
2022-01-18T13:41:26.329Z	INFO	controller.controller.volume	Starting EventSource	{"commit": "7e79a67", "reconciler group": "", "reconciler kind": "PersistentVolumeClaim", "source": "kind source: /, Kind="}
2022-01-18T13:41:26.329Z	INFO	controller.controller.volume	Starting EventSource	{"commit": "7e79a67", "reconciler group": "", "reconciler kind": "PersistentVolumeClaim", "source": "kind source: /, Kind="}
2022-01-18T13:41:26.329Z	INFO	controller.controller.volume	Starting Controller	{"commit": "7e79a67", "reconciler group": "", "reconciler kind": "PersistentVolumeClaim"}
2022-01-18T13:41:26.329Z	INFO	controller.controller.termination	Starting EventSource	{"commit": "7e79a67", "reconciler group": "", "reconciler kind": "Node", "source": "kind source: /, Kind="}
2022-01-18T13:41:26.329Z	INFO	controller.controller.termination	Starting Controller	{"commit": "7e79a67", "reconciler group": "", "reconciler kind": "Node"}
2022-01-18T13:41:26.329Z	INFO	controller.controller.node	Starting EventSource	{"commit": "7e79a67", "reconciler group": "", "reconciler kind": "Node", "source": "kind source: /, Kind="}
2022-01-18T13:41:26.329Z	INFO	controller.controller.node	Starting EventSource	{"commit": "7e79a67", "reconciler group": "", "reconciler kind": "Node", "source": "kind source: /, Kind="}
2022-01-18T13:41:26.329Z	INFO	controller.controller.node	Starting EventSource	{"commit": "7e79a67", "reconciler group": "", "reconciler kind": "Node", "source": "kind source: /, Kind="}
2022-01-18T13:41:26.329Z	INFO	controller.controller.node	Starting Controller	{"commit": "7e79a67", "reconciler group": "", "reconciler kind": "Node"}
2022-01-18T13:41:26.329Z	INFO	controller.controller.podmetrics	Starting EventSource	{"commit": "7e79a67", "reconciler group": "", "reconciler kind": "Pod", "source": "kind source: /, Kind="}
2022-01-18T13:41:26.329Z	INFO	controller.controller.podmetrics	Starting Controller	{"commit": "7e79a67", "reconciler group": "", "reconciler kind": "Pod"}
2022-01-18T13:41:26.329Z	INFO	controller.controller.nodemetrics	Starting EventSource	{"commit": "7e79a67", "reconciler group": "", "reconciler kind": "Node", "source": "kind source: /, Kind="}
2022-01-18T13:41:26.330Z	INFO	controller.controller.nodemetrics	Starting EventSource	{"commit": "7e79a67", "reconciler group": "", "reconciler kind": "Node", "source": "kind source: /, Kind="}
2022-01-18T13:41:26.330Z	INFO	controller.controller.nodemetrics	Starting EventSource	{"commit": "7e79a67", "reconciler group": "", "reconciler kind": "Node", "source": "kind source: /, Kind="}
2022-01-18T13:41:26.330Z	INFO	controller.controller.nodemetrics	Starting Controller	{"commit": "7e79a67", "reconciler group": "", "reconciler kind": "Node"}
2022-01-18T13:41:26.429Z	INFO	controller.controller.termination	Starting workers	{"commit": "7e79a67", "reconciler group": "", "reconciler kind": "Node", "worker count": 10}
2022-01-18T13:41:26.430Z	INFO	controller.controller.podmetrics	Starting workers	{"commit": "7e79a67", "reconciler group": "", "reconciler kind": "Pod", "worker count": 1}
2022-01-18T13:41:26.430Z	INFO	controller.controller.counter	Starting workers	{"commit": "7e79a67", "reconciler group": "karpenter.sh", "reconciler kind": "Provisioner", "worker count": 10}
2022-01-18T13:41:26.431Z	INFO	controller.controller.nodemetrics	Starting workers	{"commit": "7e79a67", "reconciler group": "", "reconciler kind": "Node", "worker count": 1}
2022-01-18T13:41:26.431Z	INFO	controller.controller.node	Starting workers	{"commit": "7e79a67", "reconciler group": "", "reconciler kind": "Node", "worker count": 10}
2022-01-18T13:41:26.448Z	INFO	controller.controller.volume	Starting workers	{"commit": "7e79a67", "reconciler group": "", "reconciler kind": "PersistentVolumeClaim", "worker count": 1}
2022-01-18T13:41:26.448Z	INFO	controller.controller.provisioning	Starting workers	{"commit": "7e79a67", "reconciler group": "karpenter.sh", "reconciler kind": "Provisioner", "worker count": 10}
```

**webhook**

```
2022/01/18 13:39:03 Registering 2 clients
2022/01/18 13:39:03 Registering 2 informer factories
2022/01/18 13:39:03 Registering 3 informers
2022/01/18 13:39:03 Registering 4 controllers
2022-01-18T13:39:03.746Z	INFO	Successfully created the logger.
2022-01-18T13:39:03.747Z	INFO	Logging level set to: info
2022-01-18T13:39:03.747Z	INFO	webhook	Profiling enabled: false	{"commit": "7e79a67"}
2022-01-18T13:39:03.754Z	INFO	webhook	Running with Standard leader election	{"commit": "7e79a67"}
2022-01-18T13:39:03.769Z	INFO	webhook	Starting configuration manager...	{"commit": "7e79a67"}
{"level":"info","ts":1642513143.869729,"logger":"fallback","caller":"injection/injection.go:61","msg":"Starting informers..."}
I0118 13:39:03.971345       1 leaderelection.go:243] attempting to acquire leader lease karpenter/webhook.configmapwebhook.00-of-01...
I0118 13:39:03.972109       1 leaderelection.go:243] attempting to acquire leader lease karpenter/webhook.webhookcertificates.00-of-01...
I0118 13:39:03.972415       1 leaderelection.go:243] attempting to acquire leader lease karpenter/webhook.defaultingwebhook.00-of-01...
2022-01-18T13:39:03.970Z	INFO	webhook	Informers have been synced, unblocking admission webhooks.	{"commit": "7e79a67"}
2022-01-18T13:39:03.970Z	INFO	webhook	Starting controllers...	{"commit": "7e79a67"}
2022-01-18T13:39:03.970Z	INFO	webhook	webhook.configmapwebhook.00-of-01 will run in leader-elected mode with id "karpenter-webhook-7fc64bdc8f-r664x_0e90baa0-7296-4286-b779-48cb67820bc1"	{"commit": "7e79a67"}
2022-01-18T13:39:03.970Z	INFO	webhook.ConfigMapWebhook	Starting controller and workers	{"commit": "7e79a67"}
2022-01-18T13:39:03.970Z	INFO	webhook.ConfigMapWebhook	Started workers	{"commit": "7e79a67"}
2022-01-18T13:39:03.970Z	INFO	webhook	webhook.webhookcertificates.00-of-01 will run in leader-elected mode with id "karpenter-webhook-7fc64bdc8f-r664x_0d36b46f-e71c-4e55-b027-400ebf49d74e"	{"commit": "7e79a67"}
2022-01-18T13:39:03.970Z	INFO	webhook.WebhookCertificates	Starting controller and workers	{"commit": "7e79a67"}
2022-01-18T13:39:03.970Z	INFO	webhook.WebhookCertificates	Started workers	{"commit": "7e79a67"}
2022-01-18T13:39:03.971Z	INFO	webhook	webhook.defaultingwebhook.00-of-01 will run in leader-elected mode with id "karpenter-webhook-7fc64bdc8f-r664x_249185db-f576-43dc-a48d-1a55d1c17764"	{"commit": "7e79a67"}
2022-01-18T13:39:03.971Z	INFO	webhook.DefaultingWebhook	Starting controller and workers	{"commit": "7e79a67"}
2022-01-18T13:39:03.971Z	INFO	webhook.DefaultingWebhook	Started workers	{"commit": "7e79a67"}
2022-01-18T13:39:03.971Z	INFO	webhook	webhook.validationwebhook.00-of-01 will run in leader-elected mode with id "karpenter-webhook-7fc64bdc8f-r664x_3fe7f975-962c-4654-a637-4d9c782094ed"	{"commit": "7e79a67"}
2022-01-18T13:39:03.971Z	INFO	webhook.ValidationWebhook	Starting controller and workers	{"commit": "7e79a67"}
2022-01-18T13:39:03.971Z	INFO	webhook.ValidationWebhook	Started workers	{"commit": "7e79a67"}
I0118 13:39:03.980184       1 leaderelection.go:243] attempting to acquire leader lease karpenter/webhook.validationwebhook.00-of-01...
I0118 13:39:04.030197       1 leaderelection.go:253] successfully acquired lease karpenter/webhook.defaultingwebhook.00-of-01
I0118 13:39:04.030358       1 leaderelection.go:253] successfully acquired lease karpenter/webhook.validationwebhook.00-of-01
I0118 13:39:04.030593       1 leaderelection.go:253] successfully acquired lease karpenter/webhook.configmapwebhook.00-of-01
2022-01-18T13:39:04.030Z	INFO	webhook	"karpenter-webhook-7fc64bdc8f-r664x_249185db-f576-43dc-a48d-1a55d1c17764" has started leading "webhook.defaultingwebhook.00-of-01"	{"commit": "7e79a67"}
2022-01-18T13:39:04.030Z	INFO	webhook	"karpenter-webhook-7fc64bdc8f-r664x_3fe7f975-962c-4654-a637-4d9c782094ed" has started leading "webhook.validationwebhook.00-of-01"	{"commit": "7e79a67"}
2022-01-18T13:39:04.031Z	INFO	webhook	"karpenter-webhook-7fc64bdc8f-r664x_0e90baa0-7296-4286-b779-48cb67820bc1" has started leading "webhook.configmapwebhook.00-of-01"	{"commit": "7e79a67"}
2022-01-18T13:39:04.048Z	INFO	webhook.ConfigMapWebhook	Webhook is valid	{"commit": "7e79a67", "knative.dev/traceid": "15f177f7-25ed-477c-b9b9-19e82f3de3ad", "knative.dev/key": "validation.webhook.config.karpenter.sh"}
2022-01-18T13:39:04.048Z	INFO	webhook.ConfigMapWebhook	Reconcile succeeded	{"commit": "7e79a67", "knative.dev/traceid": "15f177f7-25ed-477c-b9b9-19e82f3de3ad", "knative.dev/key": "validation.webhook.config.karpenter.sh", "duration": "17.131416ms"}
I0118 13:39:04.129683       1 leaderelection.go:253] successfully acquired lease karpenter/webhook.webhookcertificates.00-of-01
2022-01-18T13:39:04.137Z	INFO	webhook.ValidationWebhook	Updating webhook	{"commit": "7e79a67", "knative.dev/traceid": "66c338d9-2c1c-4504-b178-7d5629121289", "knative.dev/key": "validation.webhook.provisioners.karpenter.sh"}
2022-01-18T13:39:04.229Z	INFO	webhook	"karpenter-webhook-7fc64bdc8f-r664x_0d36b46f-e71c-4e55-b027-400ebf49d74e" has started leading "webhook.webhookcertificates.00-of-01"	{"commit": "7e79a67"}
2022-01-18T13:39:04.229Z	INFO	webhook.WebhookCertificates	Reconcile succeeded	{"commit": "7e79a67", "knative.dev/traceid": "59079d50-5a19-4e2b-8b81-4131c252288c", "knative.dev/key": "karpenter/karpenter-webhook-cert", "duration": "211.463µs"}
2022-01-18T13:39:04.232Z	INFO	webhook.DefaultingWebhook	Updating webhook	{"commit": "7e79a67", "knative.dev/traceid": "6475bf97-0d81-4485-ac32-16a670e66bb4", "knative.dev/key": "defaulting.webhook.provisioners.karpenter.sh"}
2022-01-18T13:39:04.233Z	INFO	webhook.ValidationWebhook	Reconcile succeeded	{"commit": "7e79a67", "knative.dev/traceid": "66c338d9-2c1c-4504-b178-7d5629121289", "knative.dev/key": "validation.webhook.provisioners.karpenter.sh", "duration": "202.937674ms"}
2022-01-18T13:39:04.243Z	INFO	webhook.DefaultingWebhook	Reconcile succeeded	{"commit": "7e79a67", "knative.dev/traceid": "6475bf97-0d81-4485-ac32-16a670e66bb4", "knative.dev/key": "defaulting.webhook.provisioners.karpenter.sh", "duration": "212.482586ms"}
```

**Comment by midestefanis** [comment]
I tried to put in debug mode but it fails me too:
```
kubectl patch configmap config-logging -n karpenter --patch '{"data":{"loglevel.controller":"debug"}}'
Error from server (InternalError): Internal error occurred: failed calling webhook "validation.webhook.config.karpenter.sh": Post "https://karpenter-webhook.karpenter.svc:443/config-validation?timeout=10s" : context deadline exceeded

kubectl get configmap                                                                                 
NAME                        DATA   AGE
config-logging              1      37m
karpenter-leader-election   0      37m
kube-root-ca.crt            1      37m
```

**Comment by mikesir87** [comment]
Ok. Those logs look reasonable. Are you following the guide from the start or using a pre-existing cluster?

**Comment by midestefanis** [comment]
I created the cluster with the same module as the example, only that a more updated version (18.2.0) has already been released. https://github.com/terraform-aws-modules/terraform-aws-eks

For example, the variable/output 'worker_iam_role_name' no longer exists.

**Comment by mikesir87** [comment]
We pinned to a previous version of the module because the 18.0 release changed quite a few things (like handling of the auth_map ConfigMap) and we haven't had a chance to test and make the appropriate changes to support 18.x. I will probably be working on that this week, but until then, it might be more expedient for you to rollback to a 17.x release. 

Out of curiosity... is there anything specific in the 18.x releases you need support for?

**Comment by midestefanis** [comment]
For now only with karpenter it seems to have broken something and I can't find what. The rest I was able to make the modifications successfully.

**Comment by midestefanis** [comment]
I have tried again, reinstalling the cluster from scratch with version 18.2.2 of the EKS module and I still have errors, only now I see other errors:

The controller pod has no errors, the webhook pod does. It seems to be a certificate problem.

**webhook pod, error logs:**
```
2022/01/24 14:44:59 http: TLS handshake error from ***:38610: tls: no certificates configured
2022-01-24T14:44:59.676Z	ERROR	webhook.ConfigMapWebhook	Reconcile error	{"commit": "723b1b7", "knative.dev/traceid": "aee311f4-57ba-49b3-8fc8-bcbe0160c328", "knative.dev/key": "validation.webhook.config.karpenter.sh", "duration": "22.585µs", "error": "secret \"karpenter-webhook-cert\" is missing \"ca-cert.pem\" key"}
2022-01-24T14:44:59.679Z	ERROR	webhook.ValidationWebhook	Reconcile error	{"commit": "723b1b7", "knative.dev/traceid": "08992e82-8f01-40d4-bf17-34e47dab42b9", "knative.dev/key": "validation.webhook.provisioners.karpenter.sh", "duration": "14.827µs", "error": "secret \"karpenter-webhook-cert\" is missing \"ca-cert.pem\" key"}
2022-01-24T14:44:59.679Z	ERROR	webhook.DefaultingWebhook	Reconcile error	{"commit": "723b1b7", "knative.dev/traceid": "878cbe53-75e1-40b0-b50d-43fb1cfe087b", "knative.dev/key": "defaulting.webhook.provisioners.karpenter.sh", "duration": "8.379µs", "error": "secret \"karpenter-webhook-cert\" is missing \"ca-cert.pem\" key"}
2022-01-24T14:44:59.681Z	ERROR	webhook.ConfigMapWebhook	Reconcile error	{"commit": "723b1b7", "knative.dev/traceid": "f4bfc1aa-cf36-473d-9fe9-0f123827f3ee", "knative.dev/key": "validation.webhook.config.karpenter.sh", "duration": "15.874µs", "error": "secret \"karpenter-webhook-cert\" is missing \"ca-cert.pem\" key"}
2022-01-24T14:44:59.684Z	ERROR	webhook.DefaultingWebhook	Reconcile error	{"commit": "723b1b7", "knative.dev/traceid": "f04a6f4e-8a4d-4622-8303-1f8ff5b106fa", "knative.dev/key": "defaulting.webhook.provisioners.karpenter.sh", "duration": "15.431µs", "error": "secret \"karpenter-webhook-cert\" is missing \"ca-cert.pem\" key"}
2022-01-24T14:44:59.684Z	ERROR	webhook.ValidationWebhook	Reconcile error	{"commit": "723b1b7", "knative.dev/traceid": "6a5cd12c-219a-419b-b17c-f704b45ebc62", "knative.dev/key": "validation.webhook.provisioners.karpenter.sh", "duration": "8.684µs", "error": "secret \"karpenter-webhook-cert\" is missing \"ca-cert.pem\" key"}
2022-01-24T14:44:59.841Z	ERROR	webhook.ConfigMapWebhook	Reconcile error	{"commit": "723b1b7", "knative.dev/traceid": "aad71be9-fc7a-4799-a865-7877690208db", "knative.dev/key": "validation.webhook.config.karpenter.sh", "duration": "149.32556ms", "error": "failed to update webhook: Operation cannot be fulfilled on validatingwebhookconfigurations.admissionregistration.k8s.io \"validation.webhook.config.karpenter.sh\": the object has been modified; please apply your changes to the latest version and try again"}
2022-01-24T14:44:59.847Z	ERROR	webhook.DefaultingWebhook	Reconcile error	{"commit": "723b1b7", "knative.dev/traceid": "6dacf741-217f-4d65-92dc-f2f0a45f350c", "knative.dev/key": "defaulting.webhook.provisioners.karpenter.sh", "duration": "110.787805ms", "error": "failed to update webhook: Operation cannot be fulfilled on mutatingwebhookconfigurations.admissionregistration.k8s.io \"defaulting.webhook.provisioners.karpenter.sh\": the object has been modified; please apply your changes to the latest version and try again"}
2022-01-24T14:44:59.939Z	ERROR	webhook.ValidationWebhook	Reconcile error	{"commit": "723b1b7", "knative.dev/traceid": "ec0cad50-22ef-4fb7-847b-60a79ea481a7", "knative.dev/key": "validation.webhook.provisioners.karpenter.sh", "duration": "203.81272ms", "error": "failed to update webhook: Operation cannot be fulfilled on validatingwebhookconfigurations.admissionregistration.k8s.io \"validation.webhook.provisioners.karpenter.sh\": the object has been modified; please apply your changes to the latest version and try again"}
```

**webhook pod, full logs:**

```
2022/01/24 14:44:59 Registering 2 clients
2022/01/24 14:44:59 Registering 2 informer factories
2022/01/24 14:44:59 Registering 3 informers
2022/01/24 14:44:59 Registering 4 controllers
2022-01-24T14:44:59.450Z	INFO	Successfully created the logger.
2022-01-24T14:44:59.450Z	INFO	Logging level set to: info
2022-01-24T14:44:59.450Z	INFO	webhook	Profiling enabled: false	{"commit": "723b1b7"}
2022-01-24T14:44:59.455Z	INFO	webhook	Running with Standard leader election	{"commit": "723b1b7"}
2022-01-24T14:44:59.463Z	INFO	webhook	Starting configuration manager...	{"commit": "723b1b7"}
{"level":"info","ts":1643035499.563796,"logger":"fallback","caller":"injection/injection.go:61","msg":"Starting informers..."}
2022/01/24 14:44:59 http: TLS handshake error from ****:38610: tls: no certificates configured
2022-01-24T14:44:59.636Z	WARN	webhook	server key missing	{"commit": "723b1b7"}
2022-01-24T14:44:59.664Z	INFO	webhook	Informers have been synced, unblocking admission webhooks.	{"commit": "723b1b7"}
2022-01-24T14:44:59.664Z	INFO	webhook	Starting controllers...	{"commit": "723b1b7"}
2022-01-24T14:44:59.665Z	INFO	webhook	webhook.configmapwebhook.00-of-01 will run in leader-elected mode with id "karpenter-webhook-75b875d699-qhzfv_c6770826-d516-4231-91c2-f976ac624e2b"	{"commit": "723b1b7"}
2022-01-24T14:44:59.665Z	INFO	webhook.ConfigMapWebhook	Starting controller and workers	{"commit": "723b1b7"}
2022-01-24T14:44:59.665Z	INFO	webhook.ConfigMapWebhook	Started workers	{"commit": "723b1b7"}
2022-01-24T14:44:59.665Z	INFO	webhook	webhook.webhookcertificates.00-of-01 will run in leader-elected mode with id "karpenter-webhook-75b875d699-qhzfv_19388874-429d-4d1e-9bd2-ec4f6c5fb3dc"	{"commit": "723b1b7"}
2022-01-24T14:44:59.665Z	INFO	webhook.WebhookCertificates	Starting controller and workers	{"commit": "723b1b7"}
2022-01-24T14:44:59.665Z	INFO	webhook.WebhookCertificates	Started workers	{"commit": "723b1b7"}
2022-01-24T14:44:59.665Z	INFO	webhook	webhook.defaultingwebhook.00-of-01 will run in leader-elected mode with id "karpenter-webhook-75b875d699-qhzfv_bb120a17-a7ee-4098-85ce-1dfcb2c7b263"	{"commit": "723b1b7"}
2022-01-24T14:44:59.665Z	INFO	webhook.DefaultingWebhook	Starting controller and workers	{"commit": "723b1b7"}
2022-01-24T14:44:59.665Z	INFO	webhook.DefaultingWebhook	Started workers	{"commit": "723b1b7"}
2022-01-24T14:44:59.665Z	INFO	webhook	webhook.validationwebhook.00-of-01 will run in leader-elected mode with id "karpenter-webhook-75b875d699-qhzfv_774ce24e-d03b-41f9-b97b-bcfecb71cc0c"	{"commit": "723b1b7"}
2022-01-24T14:44:59.665Z	INFO	webhook.ValidationWebhook	Starting controller and workers	{"commit": "723b1b7"}
2022-01-24T14:44:59.665Z	INFO	webhook.ValidationWebhook	Started workers	{"commit": "723b1b7"}
I0124 14:44:59.665820       1 leaderelection.go:243] attempting to acquire leader lease karpenter/webhook.configmapwebhook.00-of-01...
I0124 14:44:59.666160       1 leaderelection.go:243] attempting to acquire leader lease karpenter/webhook.webhookcertificates.00-of-01...
I0124 14:44:59.666305       1 leaderelection.go:243] attempting to acquire leader lease karpenter/webhook.defaultingwebhook.00-of-01...
I0124 14:44:59.666420       1 leaderelection.go:243] attempting to acquire leader lease karpenter/webhook.validationwebhook.00-of-01...
I0124 14:44:59.675819       1 leaderelection.go:253] successfully acquired lease karpenter/webhook.configmapwebhook.00-of-01
2022-01-24T14:44:59.676Z	INFO	webhook	"karpenter-webhook-75b875d699-qhzfv_c6770826-d516-4231-91c2-f976ac624e2b" has started leading "webhook.configmapwebhook.00-of-01"	{"commit": "723b1b7"}
2022-01-24T14:44:59.676Z	ERROR	webhook.ConfigMapWebhook	Reconcile error	{"commit": "723b1b7", "knative.dev/traceid": "aee311f4-57ba-49b3-8fc8-bcbe0160c328", "knative.dev/key": "validation.webhook.config.karpenter.sh", "duration": "22.585µs", "error": "secret \"karpenter-webhook-cert\" is missing \"ca-cert.pem\" key"}
I0124 14:44:59.678743       1 leaderelection.go:253] successfully acquired lease karpenter/webhook.validationwebhook.00-of-01
I0124 14:44:59.678918       1 leaderelection.go:253] successfully acquired lease karpenter/webhook.defaultingwebhook.00-of-01
I0124 14:44:59.679126       1 leaderelection.go:253] successfully acquired lease karpenter/webhook.webhookcertificates.00-of-01
2022-01-24T14:44:59.679Z	INFO	webhook	"karpenter-webhook-75b875d699-qhzfv_774ce24e-d03b-41f9-b97b-bcfecb71cc0c" has started leading "webhook.validationwebhook.00-of-01"	{"commit": "723b1b7"}
2022-01-24T14:44:59.679Z	ERROR	webhook.ValidationWebhook	Reconcile error	{"commit": "723b1b7", "knative.dev/traceid": "08992e82-8f01-40d4-bf17-34e47dab42b9", "knative.dev/key": "validation.webhook.provisioners.karpenter.sh", "duration": "14.827µs", "error": "secret \"karpenter-webhook-cert\" is missing \"ca-cert.pem\" key"}
2022-01-24T14:44:59.679Z	INFO	webhook	"karpenter-webhook-75b875d699-qhzfv_bb120a17-a7ee-4098-85ce-1dfcb2c7b263" has started leading "webhook.defaultingwebhook.00-of-01"	{"commit": "723b1b7"}
2022-01-24T14:44:59.679Z	ERROR	webhook.DefaultingWebhook	Reconcile error	{"commit": "723b1b7", "knative.dev/traceid": "878cbe53-75e1-40b0-b50d-43fb1cfe087b", "knative.dev/key": "defaulting.webhook.provisioners.karpenter.sh", "duration": "8.379µs", "error": "secret \"karpenter-webhook-cert\" is missing \"ca-cert.pem\" key"}
2022-01-24T14:44:59.679Z	INFO	webhook	"karpenter-webhook-75b875d699-qhzfv_19388874-429d-4d1e-9bd2-ec4f6c5fb3dc" has started leading "webhook.webhookcertificates.00-of-01"	{"commit": "723b1b7"}
2022-01-24T14:44:59.679Z	INFO	webhook.WebhookCertificates	Certificate secret "karpenter-webhook-cert" is missing key "server-key.pem"	{"commit": "723b1b7", "knative.dev/traceid": "6134bb0f-5a8f-412c-8eb5-ebb0aaaee161", "knative.dev/key": "karpenter/karpenter-webhook-cert"}
2022-01-24T14:44:59.681Z	ERROR	webhook.ConfigMapWebhook	Reconcile error	{"commit": "723b1b7", "knative.dev/traceid": "f4bfc1aa-cf36-473d-9fe9-0f123827f3ee", "knative.dev/key": "validation.webhook.config.karpenter.sh", "duration": "15.874µs", "error": "secret \"karpenter-webhook-cert\" is missing \"ca-cert.pem\" key"}
2022-01-24T14:44:59.684Z	ERROR	webhook.DefaultingWebhook	Reconcile error	{"commit": "723b1b7", "knative.dev/traceid": "f04a6f4e-8a4d-4622-8303-1f8ff5b106fa", "knative.dev/key": "defaulting.webhook.provisioners.karpenter.sh", "duration": "15.431µs", "error": "secret \"karpenter-webhook-cert\" is missing \"ca-cert.pem\" key"}
2022-01-24T14:44:59.684Z	ERROR	webhook.ValidationWebhook	Reconcile error	{"commit": "723b1b7", "knative.dev/traceid": "6a5cd12c-219a-419b-b17c-f704b45ebc62", "knative.dev/key": "validation.webhook.provisioners.karpenter.sh", "duration": "8.684µs", "error": "secret \"karpenter-webhook-cert\" is missing \"ca-cert.pem\" key"}
2022-01-24T14:44:59.690Z	INFO	webhook.WebhookCertificates	Reconcile succeeded	{"commit": "723b1b7", "knative.dev/traceid": "6134bb0f-5a8f-412c-8eb5-ebb0aaaee161", "knative.dev/key": "karpenter/karpenter-webhook-cert", "duration": "10.590178ms"}
2022-01-24T14:44:59.690Z	INFO	webhook.WebhookCertificates	Reconcile succeeded	{"commit": "723b1b7", "knative.dev/traceid": "adb61b15-5196-4462-9e87-40403099c3d0", "knative.dev/key": "karpenter/karpenter-webhook-cert", "duration": "142.751µs"}
2022-01-24T14:44:59.739Z	INFO	webhook.ConfigMapWebhook	Updating webhook	{"commit": "723b1b7", "knative.dev/traceid": "8918f170-09cc-47df-a72f-6bdf057c1cf7", "knative.dev/key": "karpenter/karpenter-webhook-cert"}
2022-01-24T14:44:59.741Z	INFO	webhook.ValidationWebhook	Updating webhook	{"commit": "723b1b7", "knative.dev/traceid": "cdb5895e-e3b4-4d98-8f26-fc25bf04f1b1", "knative.dev/key": "karpenter/karpenter-webhook-cert"}
2022-01-24T14:44:59.744Z	INFO	webhook.DefaultingWebhook	Updating webhook	{"commit": "723b1b7", "knative.dev/traceid": "e516ab58-85d0-4a3d-88e1-903556c58423", "knative.dev/key": "karpenter/karpenter-webhook-cert"}
2022-01-24T14:44:59.746Z	INFO	webhook.ConfigMapWebhook	Updating webhook	{"commit": "723b1b7", "knative.dev/traceid": "aad71be9-fc7a-4799-a865-7877690208db", "knative.dev/key": "validation.webhook.config.karpenter.sh"}
2022-01-24T14:44:59.838Z	INFO	webhook.ValidationWebhook	Updating webhook	{"commit": "723b1b7", "knative.dev/traceid": "ec0cad50-22ef-4fb7-847b-60a79ea481a7", "knative.dev/key": "validation.webhook.provisioners.karpenter.sh"}
2022-01-24T14:44:59.838Z	INFO	webhook.DefaultingWebhook	Updating webhook	{"commit": "723b1b7", "knative.dev/traceid": "6dacf741-217f-4d65-92dc-f2f0a45f350c", "knative.dev/key": "defaulting.webhook.provisioners.karpenter.sh"}
2022-01-24T14:44:59.840Z	INFO	webhook.ValidationWebhook	Reconcile succeeded	{"commit": "723b1b7", "knative.dev/traceid": "cdb5895e-e3b4-4d98-8f26-fc25bf04f1b1", "knative.dev/key": "karpenter/karpenter-webhook-cert", "duration": "149.09629ms"}
2022-01-24T14:44:59.841Z	INFO	webhook.ConfigMapWebhook	Reconcile succeeded	{"commit": "723b1b7", "knative.dev/traceid": "8918f170-09cc-47df-a72f-6bdf057c1cf7", "knative.dev/key": "karpenter/karpenter-webhook-cert", "duration": "150.798064ms"}
2022-01-24T14:44:59.841Z	INFO	webhook.DefaultingWebhook	Reconcile succeeded	{"commit": "723b1b7", "knative.dev/traceid": "e516ab58-85d0-4a3d-88e1-903556c58423", "knative.dev/key": "karpenter/karpenter-webhook-cert", "duration": "150.583509ms"}
2022-01-24T14:44:59.841Z	ERROR	webhook.ConfigMapWebhook	Reconcile error	{"commit": "723b1b7", "knative.dev/traceid": "aad71be9-fc7a-4799-a865-7877690208db", "knative.dev/key": "validation.webhook.config.karpenter.sh", "duration": "149.32556ms", "error": "failed to update webhook: Operation cannot be fulfilled on validatingwebhookconfigurations.admissionregistration.k8s.io \"validation.webhook.config.karpenter.sh\": the object has been modified; please apply your changes to the latest version and try again"}
2022-01-24T14:44:59.847Z	ERROR	webhook.DefaultingWebhook	Reconcile error	{"commit": "723b1b7", "knative.dev/traceid": "6dacf741-217f-4d65-92dc-f2f0a45f350c", "knative.dev/key": "defaulting.webhook.provisioners.karpenter.sh", "duration": "110.787805ms", "error": "failed to update webhook: Operation cannot be fulfilled on mutatingwebhookconfigurations.admissionregistration.k8s.io \"defaulting.webhook.provisioners.karpenter.sh\": the object has been modified; please apply your changes to the latest version and try again"}
2022-01-24T14:44:59.939Z	INFO	webhook.ConfigMapWebhook	Webhook is valid	{"commit": "723b1b7", "knative.dev/traceid": "6060dfac-1bf9-48aa-9027-5cdc0cb5b454", "knative.dev/key": "validation.webhook.config.karpenter.sh"}
2022-01-24T14:44:59.939Z	INFO	webhook.ConfigMapWebhook	Reconcile succeeded	{"commit": "723b1b7", "knative.dev/traceid": "6060dfac-1bf9-48aa-9027-5cdc0cb5b454", "knative.dev/key": "validation.webhook.config.karpenter.sh", "duration": "97.764487ms"}
2022-01-24T14:44:59.939Z	ERROR	webhook.ValidationWebhook	Reconcile error	{"commit": "723b1b7", "knative.dev/traceid": "ec0cad50-22ef-4fb7-847b-60a79ea481a7", "knative.dev/key": "validation.webhook.provisioners.karpenter.sh", "duration": "203.81272ms", "error": "failed to update webhook: Operation cannot be fulfilled on validatingwebhookconfigurations.admissionregistration.k8s.io \"validation.webhook.provisioners.karpenter.sh\": the object has been modified; please apply your changes to the latest version and try again"}
2022-01-24T14:45:00.041Z	INFO	webhook.ConfigMapWebhook	Webhook is valid	{"commit": "723b1b7", "knative.dev/traceid": "8d3e350b-9b42-4fb5-81ec-f919a0969c5f", "knative.dev/key": "validation.webhook.config.karpenter.sh"}
2022-01-24T14:45:00.041Z	INFO	webhook.ConfigMapWebhook	Reconcile succeeded	{"commit": "723b1b7", "knative.dev/traceid": "8d3e350b-9b42-4fb5-81ec-f919a0969c5f", "knative.dev/key": "validation.webhook.config.karpenter.sh", "duration": "101.952435ms"}
2022-01-24T14:45:00.136Z	INFO	webhook.DefaultingWebhook	Updating webhook	{"commit": "723b1b7", "knative.dev/traceid": "c35fe8c4-58ae-45d1-89c6-728c23e597d5", "knative.dev/key": "defaulting.webhook.provisioners.karpenter.sh"}
2022-01-24T14:45:00.139Z	INFO	webhook.ValidationWebhook	Updating webhook	{"commit": "723b1b7", "knative.dev/traceid": "64125843-b848-402e-8c36-d3f2f53d02ae", "knative.dev/key": "validation.webhook.provisioners.karpenter.sh"}
2022-01-24T14:45:00.143Z	INFO	webhook.DefaultingWebhook	Reconcile succeeded	{"commit": "723b1b7", "knative.dev/traceid": "c35fe8c4-58ae-45d1-89c6-728c23e597d5", "knative.dev/key": "defaulting.webhook.provisioners.karpenter.sh", "duration": "296.365429ms"}
2022-01-24T14:45:00.155Z	INFO	webhook.DefaultingWebhook	Updating webhook	{"commit": "723b1b7", "knative.dev/traceid": "a8dcf6fd-8dbe-4fc3-a3cc-36012849bb96", "knative.dev/key": "defaulting.webhook.provisioners.karpenter.sh"}
2022-01-24T14:45:00.236Z	INFO	webhook.ValidationWebhook	Reconcile succeeded	{"commit": "723b1b7", "knative.dev/traceid": "64125843-b848-402e-8c36-d3f2f53d02ae", "knative.dev/key": "validation.webhook.provisioners.karpenter.sh", "duration": "296.562729ms"}
2022-01-24T14:45:00.244Z	INFO	webhook.DefaultingWebhook	Reconcile succeeded	{"commit": "723b1b7", "knative.dev/traceid": "a8dcf6fd-8dbe-4fc3-a3cc-36012849bb96", "knative.dev/key": "defaulting.webhook.provisioners.karpenter.sh", "duration": "100.723965ms"}
2022-01-24T14:45:00.250Z	INFO	webhook.ValidationWebhook	Updating webhook	{"commit": "723b1b7", "knative.dev/traceid": "940093f3-dbf0-4b90-a71f-24db9b8ed2c2", "knative.dev/key": "validation.webhook.provisioners.karpenter.sh"}
2022-01-24T14:45:00.257Z	INFO	webhook.ValidationWebhook	Reconcile succeeded	{"commit": "723b1b7", "knative.dev/traceid": "940093f3-dbf0-4b90-a71f-24db9b8ed2c2", "knative.dev/key": "validation.webhook.provisioners.karpenter.sh", "duration": "21.103253ms"}
```

**Comment by ellistarn** [comment]
Does the webhook eventually recover?

**Comment by midestefanis** [comment]
No, it never recorvers. From the logs, it seems that is looking for a configmap named 'karpenter-webhook-cert' that doesn't exist.

```
kubectl get cm -n karpenter                        
NAME                        DATA   AGE
config-logging              1      3h25m
karpenter-leader-election   0      3h24m
kube-root-ca.crt            1      3h25m
```

**Comment by midestefanis** [comment]
It seems not. It keeps throwing TLS handshake error from ***:36928: EOF. The secret 'karpenter-webhook-cert' exists and has the cert so I don't understand what the problem is.

**Comment by nicolas-porta** [comment]
Hello! Any news on this ? I’m having exactly the same problemas as @midestefanis

**Comment by ellistarn** [comment]
This is going to be a challenging one, as I expect it's within knative/pkg's certificate generation logic. We're going to need to collect evidence, and hopefully eventually land on reproduction steps. In the short term, I recommend reinstalling the helm chart and see if you can move past it -- if not, then the silver lining is we have a reproduction.

**Comment by olemarkus** [comment]
Is it possible to tell karpenter not to use the knative TLS provisioning and rely on e.g ca-injection from cert-manager (e.g load balancer controller does this)?

**Comment by ellistarn** [comment]
You can use cert-manager to create/maintain this secret instead.

**Comment by midestefanis** [comment]
@ellistarn It didn't work, deleting the secret worsens the logs because that secret with its cert is not regenerated if it is deleted.

**Comment by alekc** [comment]
Are you sure that it's not a security group issue? One of the most annoying "improvement" terraform-eks-module brought in v18 is that by default nodes cannot speak one with another because Sg are very restrictive (see https://github.com/terraform-aws-modules/terraform-aws-eks/issues/1748), especially https://github.com/terraform-aws-modules/terraform-aws-eks/issues/1748#issuecomment-1020226274

I would suggest to try with v17 as suggested in the docs first if possible.

**Comment by midestefanis** [comment]
@alekc It was that! Thanks a lot!

**Comment by ellistarn** [comment]
Nice find! Would you be willing to add this to the troubleshooting guide?

**Comment by alekc** [comment]
@ellistarn will do

**Comment by prashantbhrgv** [comment]
> @alekc It was that! Thanks a lot!

I'm getting the same issue. I'm using 18.7.2 version terraform-aws-modules/eks/aws.

I tried adding the following block for the security group after following the links mentioned above

`# Extend cluster security group rules
  cluster_security_group_additional_rules = {
    egress_nodes_ephemeral_ports_tcp = {
      description                = "Node all egress"
      protocol                   = "-1"
      from_port                  = 0
      to_port                    = 0
      type                       = "egress"
      source_node_security_group = true
    }
  }

  node_security_group_additional_rules = {
    ingress_self_all = {
      description = "Node to node all ports/protocols"
      protocol    = "-1"
      from_port   = 0
      to_port     = 0
      type        = "ingress"
      self        = true
    }
    egress_all = {
      description      = "Node all egress"
      protocol         = "-1"
      from_port        = 0
      to_port          = 0
      type             = "egress"
      cidr_blocks      = ["0.0.0.0/0"]
      ipv6_cidr_blocks = ["::/0"]
    }
  }`

After adding this block, I'm still getting the following error on trying to install the provisioner

**Comment by midestefanis** [comment]
@prashantbhrgv 

Try adding these:

```
  cluster_security_group_additional_rules = {
    ingress_nodes_karpenter_ports_tcp = {
      description                = "Karpenter readiness"
      protocol                   = "tcp"
      from_port                  = 8443
      to_port                    = 8443
      type                       = "ingress"
      source_node_security_group = true
    }
  }
  
  node_security_group_additional_rules = {
    aws_lb_controller_webhook = {
      description                   = "Cluster API to AWS LB Controller webhook"
      protocol                      = "all"
      from_port                     = 9443
      to_port                       = 9443
      type                          = "ingress"
      source_cluster_security_group = true
    }
  }  
```

**Comment by prashantbhrgv** [comment]
> @prashantbhrgv
> 
> Try adding these:
> 
> ```
>   cluster_security_group_additional_rules = {
>     ingress_nodes_karpenter_ports_tcp = {
>       description                = "Karpenter readiness"
>       protocol                   = "tcp"
>       from_port                  = 8443
>       to_port                    = 8443
>       type                       = "ingress"
>       source_node_security_group = true
>     }
>   }
>   
>   node_security_group_additional_rules = {
>     aws_lb_controller_webhook = {
>       description                   = "Cluster API to AWS LB Controller webhook"
>       protocol                      = "all"
>       from_port                     = 9443
>       to_port                       = 9443
>       type                          = "ingress"
>       source_cluster_security_group = true
>     }
>   }  
> ```

Thanks for sharing this. It worked 👍🏻 . Just one question. Might be naive.

How do you determine which ports/rules need to be added? in case I face a similar issue in the future.

**Comment by alekc** [comment]
@prashantbhrgv you need to add one port for every webhook you add to the cluster (or just create a blank node->node white list like it was in the past (<v18)

**Comment by chrisnegus** [comment]
The appears to be resolved with [PR #1267](https://github.com/aws/karpenter/pull/1267),
so I'm closing this issue.
Please reopen it if there is some aspect of this issue that is still unresolved.

**Comment by Skaopap** [comment]
> @prashantbhrgv
> 
> Try adding these:
> 
> ```
>   cluster_security_group_additional_rules = {
>     ingress_nodes_karpenter_ports_tcp = {
>       description                = "Karpenter readiness"
>       protocol                   = "tcp"
>       from_port                  = 8443
>       to_port                    = 8443
>       type                       = "ingress"
>       source_node_security_group = true
>     }
>   }
>   
>   node_security_group_additional_rules = {
>     aws_lb_controller_webhook = {
>       description                   = "Cluster API to AWS LB Controller webhook"
>       protocol                      = "all"
>       from_port                     = 9443
>       to_port                       = 9443
>       type                          = "ingress"
>       source_cluster_security_group = true
>     }
>   }  
> ```

Hello,
First of all thanks, it helped me figured out my problem, but important warning here when you specify *protocol   = "all"*  acccording to [security group rule doc](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/security_group_rule)

> Setting protocol = "all" or protocol = -1 with from_port and to_port will result in the EC2 API creating a security group rule with ALL PORTS OPEN. 

It works for @prashantbhrgv because it openned all ports. So without specifying "all", you have to check your webhook endpoints as @alekc said and open the right ports. Working with alb controller and karpenter everything works fine with this config for me : 

```
cluster_security_group_additional_rules = {
    ingress_nodes_karpenter_ports_tcp = {
      description                = "Karpenter readiness"
      protocol                   = "tcp"
      from_port                  = 8443
      to_port                    = 8443
      type                       = "ingress"
      source_node_security_group = true
    }
  }

  node_security_group_additional_rules = {
    ingress_allow_alb_webhook_access_from_control_plane = {
      description                   = "Allow access from control plane to webhook port of AWS load balancer controller"
      protocol                      = "tcp"
      from_port                     = 9443
      to_port                       = 9443
      type                          = "ingress"
      source_cluster_security_group = true
    }
    ingress_allow_karpenter_webhook_access_from_control_plane = {
      description                   = "Allow access from control plane to webhook port of karpenter"
      protocol                      = "tcp"
      from_port                     = 8443
      to_port                       = 8443
      type                          = "ingress"
      source_cluster_security_group = true
    }
}
```

**Comment by dani-newman** [comment]
Still relevant and still appears on both v0.19.3 and v0.20.0.

**Comment by jtgorny** [comment]
Appears v0.19.3 doesn't have a dedicated webhook container? Therefore [troubleshooting guide is incorrect here.](https://karpenter.sh/v0.19.3/troubleshooting/#failed-calling-webhook-defaultingwebhookprovisionerskarpentersh)

Anyways, verified karpenter pod is running (1/1) & verified service endpoints exist. Validated SG ingress rules as well.

`Error from server (InternalError): error when creating "./out/karpenter/provisioner.yaml": Internal error occurred: failed calling webhook "defaulting.webhook.karpenter.k8s.aws": failed to call webhook: Post "https://karpenter.karpenter.svc:443/default/karpenter.k8s.aws?timeout=10s": Address is not allowed`

**Comment by alekc** [comment]
@jtgorny thats another issue, something from within your cluster (Istio?) is blocking access to the destination. Have a look at this https://github.com/aws/karpenter/issues/1602 discussion for some cluses
